Mongo REST upload

AdrianLxM opened the pull request:
TESTING - please do NOT merge

With the mongo 3 update we will lose the ability to directly upload to mongolab as the new java-driver does not run on Android.

As some people (including me) would still prefer direct Mongolab upload to the upload via Nightscout (not using NS, using Mongolab just to mirror data or to view on Garmin watch, ...), I wrote a first version of a REST-Uploader for MongoLab. The Rest-API upload is used by the chrome uploader as well and seems to be pretty stable.
We also may use this as a starting point to rewrite the NS-Rest-Upload, as they are pretty similar (code borrowed from current Rest-Upload).

So far it worked for me and takes 19930.0 ms to upload 32 records.

For testing:
- enable mongolab in settings (url does not matter)
- in class com.eveningoutpost.dexdrip.UtilityModels.MongoLabRest:
  - enter your database name as string constant DATABASE
  - enter your apiKey as string constant API_KEY (http://docs.mongolab.com/data-api/#authentication)

It would be great if you could test that! If it works, we can move the db-name and API-Secret to the settings and clean the old code from NightscoutUploader.
(@tzachi-dar - this is a newer version than I mailed you.)

Commit:
Rest uploader for MongoLab

Commit:
Rest/Mongo fix and hooked in

StephenBlackWasAlreadyTaken commented:
This sounds like a good solution to me.
I would much prefer Mongo rest to just cramming it in to the db manually
(and if it didn't change protocols for Mongo 3 than why not! It's a fairly
easy setup)

Will give this a go later today

On Thu, Sep 24, 2015, 12:46 AM AdrianLxM notifications@github.com wrote:

> TESTING - please do NOT merge
> 
> With the mongo 3 update we will lose the ability to directly upload to
> mongolab as the new java-driver does not run on Android.
> 
> As some people (including me) would still prefer direct Mongolab upload to
> the upload via Nightscout (not using NS, using Mongolab just to mirror data
> or to view on Garmin watch, ...), I wrote a first version of a
> REST-Uploader for MongoLab. The Rest-API upload is used by the chrome
> uploader as well and seems to be pretty stable.
> We also may use this as a starting point to rewrite the NS-Rest-Upload, as
> they are pretty similar (code borrowed from current Rest-Upload).
> 
> So far it worked for me and takes 19930.0 ms to upload 32 records.
> 
> For testing:
> - enable mongolab in settings (url does not matter)
> - in class com.eveningoutpost.dexdrip.UtilityModels.MongoLabRest:
>   - enter your database name as string constant DATABASE
>   - enter your apiKey as string constant API_KEY (
>     http://docs.mongolab.com/data-api/#authentication)
> 
> It would be great if you could test that! If it works, we can move the
> db-name and API-Secret to the settings and clean the old code from
> NightscoutUploader.
> (@tzachi-dar https://github.com/tzachi-dar - this is a newer version
> 
> ## than I mailed you.)
> 
> You can view, comment on, or merge this pull request online at:
> 
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157
> Commit Summary
> - Rest uploader for MongoLab
> - Rest/Mongo fix and hooked in
> 
> File Changes
> - _A_
>   app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/MongoLabRest.java
>   https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157/files#diff-0
>   (234)
> - _M_
>   app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java
>   https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157/files#diff-1
>   (15)
> 
> Patch Links:
> 
>    -
>    https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157.patch
>    -
>    https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157
> .

tzachi-dar commented:
This also seems to me like a good step.
We will need to also have UI for this.
I'll also add a get function that will replace the get that the wifi thread was doing.

tzachi-dar commented:
We should probably also make our UI for direct mongo uploading gray, since it will stop working soon.

StephenBlackWasAlreadyTaken commented:
Someone may be rolling their own Mingo rather than using Mongo labs

On Thu, Sep 24, 2015, 8:43 AM tzachi-dar notifications@github.com wrote:

> We should probably also make our UI for direct mongo uploading gray, since
> it will stop working soon.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157#issuecomment-142917068
> .

tzachi-dar commented:
Didn't really think about creating my own mongodb. This could also have been a solution...

AdrianLxM commented:
> Someone may be rolling their own Mingo rather than using Mongo labs

For Mongo 3.0 we unfortunately won't have any solution - until a driver comes out. (Please correct me if I'm wrong)
The only thing I can think of, is make the baseURL also configurable. Then people can run their own REST server... overkill?

The ones running their own Mongo-Server also don't have to update to mongo 3.0. Version 2.6 still gets maintenance updates.
I like @tzachi-dar 's idea to grey out the "Mongo Legacy upload" but still have it as an option. People running their own mongo-server usully don't have problems configuring things in a deeper android setting.

StephenBlackWasAlreadyTaken commented:
Yeah, I 100% agree with that, calling it Mongo Legacy Upload makes sense, I also think we should call it Mongo API instead of Mongo REST (just to be less confusing while people are being forced to use rest)

tzachi-dar commented:
I don't think that changing the base url is enough for people to use other servers. for example azure have their own secret format.

StephenBlackWasAlreadyTaken commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/MongoLabRest.java:
```diff
@@ -0,0 +1,234 @@
+    // Should be read from settings/preferences in final version
+    private static final String DATABASE = "mydb";
+    private static final String API_KEY = ;
+    // How to get an API-Key: http://docs.mongolab.com/data-api/#authentication
```
This for me was the hardest part of setup :-D

AdrianLxM commented:
:+1: on "Mongo API"

StephenBlackWasAlreadyTaken commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java:
```diff
@@ -86,8 +86,19 @@ public boolean upload(List<BgReading> glucoseDataSets, List<Calibration> meterRe
 
             if (enableMongoUpload) {
```
Im an idiot and didnt realize why it wasnt uploading... (I thought you had hardcoded it to just upload no matter what woooops!)

tzachi-dar committed:
First version that reads mongo using the rest api.

Still needs cleaning.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>

tzachi-dar committed:
Read rest api parameters from the preferences.

Made the send code robuster.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>

tzachi-dar committed:
Try to increase the stability of rest reader.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>

Commit:
default collections

"entries" for meter readings, calibrations and SGV.
"devicestatus" for the devicestatus.

Commit:
entries for data

Commit:
created_at

StephenBlackWasAlreadyTaken commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/MongoLabRest.java:
```diff
@@ -0,0 +1,378 @@
+    // FOR TESTING:
+    // Should be read from settings/preferences in final version
+    private static final String DATABASE = "mydb";
+    private static final String API_KEY = "D2a6iaurh-oihXrraOquZSySx9QnT_Gs";
```
Im guessing you didnt mean to check this in

tzachi-dar replied:
sure, I'll remove it soon. this is  not used any more.

StephenBlackWasAlreadyTaken commented on app/src/main/res/xml/pref_data_sync.xml:
```diff
@@ -20,14 +20,34 @@
+                android:key="cloud_storage_mongodb_rest_key"
+                android:dialogTitle="@string/pref_dialog_mongodb_collection"
+                android:dialogMessage="This is the key that identifies your rest api"
+                android:defaultValue="D2a6iaurh-oihXrraOquZSySx9QnT_G">
```
same

tzachi-dar replied:
Here, I did mean to check it in. This is not the same string, and I wanted to give it as an example of what people will want to give.

tzachi-dar committed:
removed unneeded data.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>

AdrianLxM commented:
I'll close this pr and delete (=hide, restorable) the branch as the upload now works with the driver again.

tzachi-dar commented:
How can i find this branch?
בתאריך 18 באוק 2015 17:31,‏ "AdrianLxM" notifications@github.com כתב:

> I'll close this pr and delete (=hide, restorable) the branch as the upload
> now works with the driver again.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/157#issuecomment-149020388
> .

AdrianLxM commented:
@tzachi-dar I just restored the branch
